File: 124.cpp
#include<iostream>
#include<vector>

using namespace std;

struct TreeNode{
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int x):val(x),left(NULL),right(NULL){}
};

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxSum=INT_MIN;
        pathSum(root,maxSum);
        return maxSum;
    }
    int pathSum(TreeNode* root,int& maxSum){
        if(!root)
            return 0;
        int left=max(pathSum(root->left,maxSum),0);
        int right=max(pathSum(root->right,maxSum),0);
        maxSum=max(maxSum,left+right+root->val);
        return max(left,right)+root->val;
    }
};

int main(){

}

File: Python3/124.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.maxsum = -float("inf")

        def PathSum(root):
            if not root:
                return 0
            leftsum = max(0, PathSum(root.left))
            rightsum = max(0, PathSum(root.right))
            self.maxsum = max(self.maxsum, leftsum + rightsum + root.val)
            return max(leftsum, rightsum) + root.val
        
        PathSum(root)
        return self.maxsum
